overrides.wake

```diff
@@ -0,0 +1,30 @@
+# Overriding example, a resource where 1.0.0 is replaced with 1.0.3
+#
+#   publish resourceOverrides = Pair "foo/baa/1.0.0" "foo/baa/1.0.3", Nil
+#
+# The Job's Plan will need to use either 'setDefaultResources' instead of
+# 'setPlanResources' or to instead manually use 'resolveOverrideableResources'
+# to make the mapping.
+#
+#   makePlan ("baa", Nil) Nil
+#   | setDefaultResources ("foo/baa/1.0.0", Nil)
+#   | runJob
+
+global topic resourceOverrides: Pair String String
+
+# Add a resource to a plan that can be later swapped via the
+# 'resourceOverrides' topic.
+# If the resources aren't matches the resources are used unchanged.
+global def setDefaultResources defaultResources plan =
```

Comment:
I'm uncomfortable with this approach, and I think it's too dangerous and invasive for every rule definition to use this function when it's not a general solution. I don't see this being our long-term solution to the problem, and to force every rule that wants overridable resources to opt in to this API will just make it more difficult to remove in the future.

I also don't think it's good to define non-environment-package APIs in this repo, since in order to depend on this function, you must pull in the other global runners defined here.

Could we do something different that is much less invasive? Is it possible to define a custom runner that does the resource swapping? That way, we aren't defining a new de facto Plan API, and instead it's completely invisible to any plans that have been defined.

Reply:
I don't think I understand runners enough to see how to replace the resource string in-flight, that would need some guidance

Reply:
I feel like we should be able to add to the `Plan` api, I think we're a little constrained by it being part of the wake stdlib.

If runJob were modifiable (perhaps living in wake-utils) we could just modify that directly to do whatever necessary transparently.

Reply:
The more I think about this, the more I feel that this should be done by the runner system, since that is exactly the exposed interface for this kind of thing. The `Plan` already has an API for declaring the required (suggested?) resources, which is the `Resources` property of the `Plan`.

I think I am slightly more open to modifying `runJob`, but I think adding a new runner means that no changes need to be made to any of the rules. The best documentation we have on runners is at https://github.com/sifive/wake/blob/cc481212b03f1b5a978e7600e1183cc0c144fcf6/share/doc/wake/quickref.md#runners, but I can talk to you offline about this. I think with scoring, we can continue doing the multiplier thing or maybe just make it be some fixed value that someone sets when they declare an override.

Reply:
Its not the 'what the score should be' question for me, but how does it know to go backward and iterate though the other runners to figure out which one to use for the new resource string.

Reply:
I thought there was a way of doing that, but I think I was mistaking it with how the cascade of runner transforms works.

I don't think this is well documented, but the `pre` argument to `makeRunner` is expected to pop off any resource that it fulfills so that other runners can also have the chance of transforming the Plan. I think this is the only way you can compose two runners that do not need to know about each other.

@terpstra, can you confirm what I said above about runner composition?